src/Network/Ricochet/Connection.hs

```diff
@@ -24,53 +25,90 @@ import           Network.Socks5           (socksConnectTo)
+                   -> Bool                -- ^ Wether this side accepted the connection
+                   -> Ricochet Connection -- ^ Returns the finished 'Connection'
+initiateConnection handle isClientSide = do
+  connection <- createConnection handle isClientSide
+  if isClientSide
+    then offerVersions connection
+    else pickVersion connection
+  return connection
+
+-- | Creates a new 'Connection' and adds it to the list of open Connections
+createConnection :: Handle              -- ^ Handle corresponding to the connection to the peer
+                 -> Bool                -- ^ Wether this side accepted the connection
+                 -> Ricochet Connection -- ^ Returns the finished 'Connection'
+createConnection handle isClientSide = do
+  -- Disable buffering
   liftIO $ hSetBuffering handle NoBuffering
-  let con = MkConnection handle [MkChannel 0 $ MkChannelType "im.ricochet.control-channel"] isClientSide B.empty
-  connections %= (<> [con])
-  vers <- use versions
-  if isClientSide then do
-    liftIO . B.hPutStr handle $ dumpIntroduction vers
-    resp <- liftIO $ B.hGet handle 1
-    let choice = head $ B.unpack resp
-    if choice `M.member` vers then
-      vers M.! choice $ con
-    else liftIO $ putStrLn "Server responded with invalid protocol choice"
-  else do
-    maybeStuff <- liftIO $ awaitIntroMessage vers handle
-    -- TODO: This is not elegant, use pattern guards
-    case fmap (first M.toList) maybeStuff of
-      Just ([], rest) -> do
-        liftIO $ putStrLn "We don’t have any versions in common with remote side"
-        liftIO . B.hPutStr handle $ B.singleton 0xFF
-      Just (handlers, rest) -> do
-        let chosen = foldl1 max (fmap fst handlers)
-        liftIO . putStrLn $ "We can choose between " <> show (length handlers) <> " versions!"
-        liftIO . putStrLn $ "We have chosen " <> show chosen <> "."
-        liftIO . B.hPutStr handle $ B.singleton chosen
-        connections . traverse . filtered (== con) . cInputBuffer %= (<> rest)
-        (fromJust $ lookup chosen handlers) con
-      Nothing -> liftIO $ putStrLn "Remote side sent invalid version negotiation."
-  return con
+  -- Create the actual connection structure
+  let channels = [MkChannel 0 $ MkChannelType "im.ricochet.control-channel"]
+      connection = MkConnection handle channels isClientSide B.empty
+  -- Add it to the list of open Connections
+  connections %= (<> [connection])
+  return connection
+
+-- | Offers the available versions to the newly connected peer and starts the
+--   handler corresponding to the one the peer picked
+offerVersions :: Connection -> Ricochet ()
+offerVersions connection = do
+  availableVersions <- use versions
+  -- Send the available versions
+  liftIO . B.hPutStr (connection ^. cHandle) $ dumpIntroduction availableVersions
+  response <- liftIO $ B.hGet (connection ^. cHandle) 1
+  let choice = head $ B.unpack response
+  -- If the choice is valid
+  if choice `M.member` availableVersions
```

Comment:
We should explicitly check for the `0xFF` here since that version could be part of the `Map` if somebody used the library wrong.